pydantic/main.py

```diff
@@ -538,27 +538,27 @@ def __iter__(self) -> TupleGenerator:
         yield from self.__dict__.items()
 
     def __eq__(self, other: Any) -> bool:
-        if not isinstance(other, BaseModel):
-            return False
-
-        # When comparing instances of generic types for equality, as long as all field values are equal,
-        # only require their generic origin types to be equal, rather than exact type equality.
-        # This prevents headaches like MyGeneric(x=1) != MyGeneric[Any](x=1).
-        self_type = self.__pydantic_generic_metadata__['origin'] or self.__class__
-        other_type = other.__pydantic_generic_metadata__['origin'] or other.__class__
-
-        if self_type != other_type:
-            return False
-
-        if self.__dict__ != other.__dict__:
-            return False
+        if isinstance(other, BaseModel):
+            # When comparing instances of generic types for equality, as long as all field values are equal,
+            # only require their generic origin types to be equal, rather than exact type equality.
+            # This prevents headaches like MyGeneric(x=1) != MyGeneric[Any](x=1).
+            self_type = self.__pydantic_generic_metadata__['origin'] or self.__class__
```

Comment:
I don't know if the ordering here has any impact on performance, but it seemed to be it would at worst be equal speed for the common case of comparing two BaseModel instances, and it didn't seem to me to have much impact on readability.

Reply:
(This looks like a big diff, but the only thing that changed was the order of the if statement, and the fact that if the types are different, it now returns NotImplemented rather than False)